### PR TITLE
Implement EIP-2718 and EIP-2930 for EIP-1559

### PR DESCRIPTION
**Pull request by gzliudan**
# Proposed changes

Pick below PRs from ethereum:

- 21502: #21502
- 22456: #22456
- 22480: #22480
- 22402: #22402
- 22333: #22333
- 22550: #22550
- 27860: #27860

With extra commits:
- rename struct callmsg to callMsg

## Types of changes

What types of changes does your code introduce to XDC network?
_Put an `✅` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Regular KTLO or any of the maintaince work. e.g code style
- [ ] CICD Improvement

## Impacted Components
Which part of the codebase this PR will touch base on,

_Put an `✅` in the boxes that apply_

- [ ] Consensus
- [ ] Account
- [ ] Network
- [X] Geth
- [ ] Smart Contract
- [ ] External components
- [ ] Not sure (Please specify below)

## Checklist
_Put an `✅` in the boxes once you have confirmed below actions (or provide reasons on not doing so) that_

- [X] This PR has sufficient test coverage (unit/integration test) OR I have provided reason in the PR description for not having test coverage
- [ ] Provide an end-to-end test plan in the PR description on how to manually test it on the devnet/testnet.
- [ ] Tested the backwards compatibility.
- [ ] Tested with XDC nodes running this version co-exist with those running the previous version.
- [ ] Relevant documentation has been updated as part of this PR
- [ ] N/A

### CODE REVIEW

**Review by jrwbabylonchain — APPROVED** [review]

**Review by wjrjerome — APPROVED** [review]

File: params/config.go [hunk 1]
```diff
@@ -552,37 +560,37 @@ func (c *ChainConfig) IsConstantinople(num *big.Int) bool {
 // - equal to or greater than the PetersburgBlock fork block,
 // - OR is nil, and Constantinople is active
 func (c *ChainConfig) IsPetersburg(num *big.Int) bool {
-	return isForked(common.TIPXDCXCancellationFee, num)
+	return isForked(common.TIPXDCXCancellationFee, num) || isForked(c.PetersburgBlock, num)
```
Comment (wgr523) [inline params/config.go:563]: Shall we remove `common.XXXBlock` then?

**Review by wgr523 — APPROVED** [review]

File: params/config.go (hunk 1 above)
Comment (gzliudan) [reply params/config.go:563]: Good idea. let's do it later.

Comment (wgr523) [reply params/config.go:563]: Nvm, it can be done in another PR

> Shall we remove `common.XXXBlock` then?

File: core/types/receipt.go [hunk 2]
```diff
@@ -93,8 +110,13 @@ type receiptStorageRLP struct {
...
 
 // NewReceipt creates a barebone transaction receipt, copying the init fields.
+// Deprecated: create receipts using a struct literal instead.
 func NewReceipt(root []byte, failed bool, cumulativeGasUsed uint64) *Receipt {
-	r := &Receipt{PostState: common.CopyBytes(root), CumulativeGasUsed: cumulativeGasUsed}
+	r := &Receipt{
+		Type:              LegacyTxType,
+		PostState:         common.CopyBytes(root),
+		CumulativeGasUsed: cumulativeGasUsed,
+	}
```
Comment (liam-lai) [inline core/types/receipt.go:119]: will this impact a disk or block format? Or just api response?

Comment (gzliudan) [reply core/types/receipt.go:119]: Both of the functions `EncodeRLP` and `DecodeRLP` support the old and new formats of Receipt at same time.

**Pull request merged**